software/contrib/euclidean_rhythm.py

```diff
@@ -0,0 +1,356 @@
+                if self.notes_para[self.selection_idx][2] < 0: self.notes_para[self.selection_idx][2] = 0
+                if self.notes_para[self.selection_idx][3] > 3: self.notes_para[self.selection_idx][3] = 3
+                if self.notes_para[self.selection_idx][3] < 0: self.notes_para[self.selection_idx][3] = 0
+                #print(self.selection_idx, self.notes_para[self.selection_idx])
+            else:
+                self.clock_flg = 1
+
+    def bjorklund(self, steps, pulses):
+        steps = int(steps)
+        pulses = int(pulses)
+        
+        if pulses > steps:
+            raise ValueError    
+        
+        pattern = []    
+        counts = []
+        remainders = []
+        
+        divisor = steps - pulses
+        remainders.append(pulses)
+        level = 0
+        while True:
+            counts.append(divisor // remainders[level])
+            remainders.append(divisor % remainders[level])
+            divisor = remainders[level]
+            level = level + 1
+            if remainders[level] <= 1:
+                break
+        counts.append(divisor)
+        
+        def build(level):
+            if level == -1:
+                pattern.append(0)
+            elif level == -2:
+                pattern.append(1)         
+            else:
+                for i in range(0, counts[level]):
+                    build(level - 1)
+                if remainders[level] != 0:
+                    build(level - 2)
+        
+        build(level)
+        i = pattern.index(1)
+        pattern = pattern[i:] + pattern[0:i]
+        return pattern
+
+    # Small dot indicators shown current editing parameter
+    def draw_editing_idx(self, selection_idx, editing_idx, cv_idx):
+        if editing_idx == 0:
+            oled.fill_rect(box_position[selection_idx][0]+29, box_position[selection_idx][1]+1, 2, 2, 1)
+        if editing_idx == 1:
+            oled.fill_rect(box_position[selection_idx][0]+29, box_position[selection_idx][1]+1, 2, 2, 1)
+            oled.fill_rect(box_position[selection_idx][0]+32, box_position[selection_idx][1]+1, 2, 2, 1)
+        if editing_idx == 2:
+            oled.fill_rect(box_position[selection_idx][0]+29, box_position[selection_idx][1]+1, 2, 2, 1)
+            oled.fill_rect(box_position[selection_idx][0]+32, box_position[selection_idx][1]+1, 2, 2, 1)
+            oled.fill_rect(box_position[selection_idx][0]+35, box_position[selection_idx][1]+1, 2, 2, 1)
+        if editing_idx == 3:
+            oled.fill_rect(box_position[selection_idx][0]+29, box_position[selection_idx][1]+1, 2, 2, 1)
+            oled.fill_rect(box_position[selection_idx][0]+32, box_position[selection_idx][1]+1, 2, 2, 1)
+            oled.fill_rect(box_position[selection_idx][0]+35, box_position[selection_idx][1]+1, 2, 2, 1)
+            oled.fill_rect(box_position[selection_idx][0]+38, box_position[selection_idx][1]+1, 2, 2, 1)
```

Comment:
The if structure above can be simplified such that the line to draw each rect only needs to occur once. The first box always gets drawn. Each subsequent box gets drawn for every larger index. So I think this would work:

```python
oled.fill_rect(box_position[selection_idx][0]+29, box_position[selection_idx][1]+1, 2, 2, 1)
if editing_idx >= 1:
   oled.fill_rect(box_position[selection_idx][0]+32, box_position[selection_idx][1]+1, 2, 2, 1)
if editing_idx >= 2:
   oled.fill_rect(box_position[selection_idx][0]+35, box_position[selection_idx][1]+1, 2, 2, 1)
if editing_idx >= 3:
   oled.fill_rect(box_position[selection_idx][0]+38, box_position[selection_idx][1]+1, 2, 2, 1)
```

Besides smaller code, this has the advantage that in the event you want to change these rects you only need to change it in one place.

Reply:
Improved.